worlds/ladx/__init__.py

```diff
@@ -241,41 +243,102 @@ def create_items(self) -> None:
-        self.multiworld.random.shuffle(self.prefill_own_dungeons)
-        self.multiworld.random.shuffle(dungeon_locations)
-        fill_restrictive(self.multiworld, all_state, dungeon_locations, self.prefill_own_dungeons, lock=True)
+            # set up allow-list for dungeon specific items
+            locs = set(self.dungeon_locations_by_dungeon[dungeon_index])
+            for item in self.prefill_original_dungeon[dungeon_index]:
+                allowed_locations_by_item[item] = locs
+
+            # put the items for this dungeon in the list to fill
+            all_dungeon_items_to_fill.extend(self.prefill_original_dungeon[dungeon_index])
+
+            # ...and gather the list of all dungeon locations
+            all_dungeon_locs |= locs
+            # ...also set the rules for the dungeon
+            for location in locs:
+                orig_rule = location.item_rule
+                # If an item is about to be placed on a dungeon location, it can go there iff 
+                # 1. it fits the general rules for that location (probably 'return True' for most places)
+                # 2. Either
+                #    2a. it's not a restricted dungeon item
+                #    2b. it's a restricted dungeon item and this location is specified as allowed
+                location.item_rule = lambda item, location=location, orig_rule=orig_rule: \
+                    (item not in allowed_locations_by_item or location in allowed_locations_by_item[item]) and orig_rule(item)
+
+        # Now set up the allow-list for any-dungeon items
+        for item in self.prefill_own_dungeons:
+            # They of course get to go in any spot
+            allowed_locations_by_item[item] = all_dungeon_locs
+
+        # Get the list of locations and shuffle
+        all_dungeon_locs_to_fill = list(all_dungeon_locs)
+        self.multiworld.random.shuffle(all_dungeon_locs_to_fill)
+
+        # Get the list of items and sort by priority
+        def priority(item):
+            # 0 - Nightmare dungeon-specific
+            # 1 - Key dungeon-specific
+            # 2 - Other dungeon-specific
+            # 3 - Nightmare any local dungeon
+            # 4 - Key any local dungeon
+            # 5 - Other any local dungeon
+            i = 2
+            if "Nightmare" in item.name:
+                i = 0
+            elif "Key" in item.name:
+                i = 1
+            if allowed_locations_by_item[item] is all_dungeon_locs:
+                i += 3
+            return i
+        all_dungeon_items_to_fill.sort(key=priority)
+
+        # Set up state
+        all_state = self.multiworld.get_all_state(use_cache=False)
+        # Remove dungeon items we are about to put in from the state so that we don't double count
+        for item in all_dungeon_items_to_fill:
+            all_state.remove(item)
+            
+        # ...and flag items as pre-filled
+        self.pre_fill_items.extend(all_dungeon_items_to_fill)

```

Comment:
does this do anything? Or rather is this required?
pre fill items are meant to alert AP ahead of pre fill of what will be created during pre fill. So extending this during pre fill doesn't seem correct to me.

Reply:
fixed?

Reply:
I don't think so? The result of get_pre_fill_items wasn't meant to mutate once pre_fill is started.

Reply:
fixed for real - I implemented the new version of it and forgot to remove that line